### PR TITLE
Implement `readlink`

### PR DESCRIPTION
**Pull request by Aaron1011**
Due to the truncating behavior of `readlink`, I was not able to
directly use any of the existing C-cstring helper functions.

### CODE REVIEW

File: src/shims/os_str.rs [hunk 1]
```diff
@@ -265,4 +251,21 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+    #[cfg(not(unix))]
+    fn os_str_to_bytes<'a>(&mut self, os_str: &'a OsStr) -> InterpResult<'tcx, &'a [u8]> {
+        // On non-unix platforms the best we can do to transform bytes from/to OS strings is to do the
+        // intermediate transformation into strings. Which invalidates non-utf8 paths that are actually
+        // valid.
+        os_str
+            .to_str()
+            .map(|s| s.as_bytes())
+            .ok_or_else(|| err_unsup_format!("{:?} is not a valid utf-8 string", os_str).into())
+    }
```
Comment (RalfJung) [inline src/shims/os_str.rs:269]: Could you move those up above the functions that use them? I usually prefer functions to be order callee-to-caler.

Also for consistency, maybe move out the `bytes_to_os_str` functions as well?

File: src/shims/posix/fs.rs [hunk 2]
```diff
@@ -1355,6 +1355,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+
+        let pathname = this.read_path_from_c_str(this.read_scalar(pathname_op)?.check_init()?)?;
+        let buf = this.read_scalar(buf_op)?.check_init()?;
+        let bufsize = this.read_scalar(bufsize_op)?.to_machine_usize(this)?;
+
+        let result = std::fs::read_link(pathname);
+        match result {
+            Ok(resolved) => {
+                let mut path_bytes = this.os_str_to_bytes(resolved.as_ref())?;
+                if path_bytes.len() > bufsize as usize {
```
Comment (RalfJung) [inline src/shims/posix/fs.rs:1377]: Please avoid `as`, use checked conversions instead that panic on overflow.

File: src/shims/posix/fs.rs [hunk 3]
```diff
@@ -1355,6 +1355,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+                    path_bytes = &path_bytes[..(bufsize as usize)]
+                }
+                // 'readlink' truncates the resolved path if
+                // the provided buffer is not large enough
+                this.memory.write_bytes(buf, path_bytes.iter().copied())?;
+                Ok(path_bytes.len() as i64)
```
Comment (RalfJung) [inline src/shims/posix/fs.rs:1383]: Same here re: `as`.

File: tests/run-pass/fs.rs [hunk 4]
```diff
@@ -215,6 +221,43 @@ fn test_symlink() {
     let mut contents = Vec::new();
     symlink_file.read_to_end(&mut contents).unwrap();
     assert_eq!(bytes, contents.as_slice());
+
+
+    #[cfg(unix)]
+    {
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:227]: We already have an `fs_libc.rs` test file, wouldn't these tests fit better there?

File: tests/run-pass/fs.rs [hunk 5]
```diff
@@ -215,6 +221,43 @@ fn test_symlink() {
...
+        use std::os::unix::ffi::OsStrExt;
+
+        let expected_path = path.as_os_str().as_bytes();
+
+        // Test that the expected string gets written to a buffer of proper
+        // length, and that a trailing null byte is not written
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:233]:
```suggestion
        // length, and that a trailing null byte is not written.
```

File: tests/run-pass/fs.rs [hunk 6]
```diff
@@ -215,6 +221,43 @@ fn test_symlink() {
...
+        let symlink_c_str = CString::new(symlink_path.as_os_str().as_bytes()).unwrap();
+        let symlink_c_ptr = symlink_c_str.as_ptr();
+
+        // Make the buf one byte larger than it needs to be,
+        // and check that the last byte is not overwritten
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:238]:
```suggestion
        // and check that the last byte is not overwritten.
```

File: tests/run-pass/fs.rs [hunk 7]
```diff
@@ -215,6 +221,43 @@ fn test_symlink() {
...
+        let mut large_buf = vec![0xFF; expected_path.len() + 1];
+        let res = unsafe { libc::readlink(symlink_c_ptr, large_buf.as_mut_ptr().cast(), large_buf.len()) };
+        assert_eq!(res, large_buf.len() as isize - 1);
+        // Check that the resovled path was properly written into the buf
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:242]:
```suggestion
        // Check that the resovled path was properly written into the buf.
```

**Review by RalfJung — COMMENTED** [review]
Thanks a lot. I particularly like the extensive test. :)

File: tests/run-pass/fs.rs (hunk 4 above)
Comment (Aaron1011) [reply tests/run-pass/fs.rs:227]: This file already has some symlink setup code that I didn't want to duplicate.

Comment (RalfJung) [reply tests/run-pass/fs.rs:227]: That's fair, but then the fs.rs vs fs_libc.rs separation doesn't really hold up any more.

Comment (Aaron1011) [reply tests/run-pass/fs.rs:227]: Both of those files have the same compilation flags - should I just merge them?

File: src/shims/os_str.rs [hunk 8]
```diff
@@ -62,27 +62,28 @@ fn convert_path_separator<'a>(
 
 impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+
+    #[cfg(unix)]
+    fn bytes_to_os_str<'a>(&self, bytes: &'a [u8]) -> InterpResult<'tcx, &'a OsStr> {
```
Comment (RalfJung) [inline src/shims/os_str.rs:67]: Why did you add a `self` parameter?

File: tests/run-pass/fs.rs (hunk 4 above)
Comment (RalfJung) [reply tests/run-pass/fs.rs:228]: Yeah seems reasonable.

File: src/shims/os_str.rs (hunk 8 above)
Comment (Aaron1011) [reply src/shims/os_str.rs:67]: This makes it easier to invoke from outside this module - if it's an inherent or freestanding method, you need to name this module or this `EvalContextExt` trait or explicitly.

Comment (RalfJung) [reply src/shims/os_str.rs:67]: That's what we do for other context-independent helper functions though, so I think we should also do it here.

Comment (Aaron1011) [reply src/shims/os_str.rs]: Done

**Comment by RalfJung** [comment]
Thanks. :)
@bors r+

File: tests/run-pass/fs.rs [hunk 9]
```diff
@@ -215,6 +222,43 @@ fn test_symlink() {
...
+        // Test that the expected string gets written to a buffer of proper
+        // length, and that a trailing null byte is not written.
+        let symlink_c_str = CString::new(symlink_path.as_os_str().as_bytes()).unwrap();
+        let symlink_c_ptr = symlink_c_str.as_ptr();
+
+        // Make the buf one byte larger than it needs to be,
+        // and check that the last byte is not overwritten.
+        let mut large_buf = vec![0xFF; expected_path.len() + 1];
+        let res = unsafe { libc::readlink(symlink_c_ptr, large_buf.as_mut_ptr().cast(), large_buf.len()) };
+        assert_eq!(res, large_buf.len() as isize - 1);
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:242]: Looks like this assertion is failing on a Windows host when running the Linux target:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `67`,
 right: `68`', $DIR/fs.rs:242:9
```

**Comment by Aaron1011** [comment]
@RalfJung: I've swapped the order of some of the assertions so that the actual paths will be compared first. Can we re-run the tests on windows?

**Comment by RalfJung** [comment]
Sure: @bors try

You can now also do this yourself: @bors delegate
Just please don't r+ the PR :)

**Comment by RalfJung** [comment]
Hm maybe I cannot have multiple bors commands in one comment?

@bors delegate

**Comment by Aaron1011** [comment]
Left path:  `C:\Users\appveyor\AppData\Local\Temp\1\miri_test_fs_link_target.txt`
Right path: `C:\Users\appveyor\AppData\Local\Temp\1\/miri_test_fs_link_target.txt`

It looks like we're joining the path with the wrong platform separator somewhere.

**Comment by Aaron1011** [comment]
Oh, I guess that's why those helper functons exist in the first place :laughing: 
I'll add in a truncation flag.

File: src/shims/os_str.rs [hunk 10]
```diff
@@ -14,52 +14,11 @@ use rustc_target::abi::LayoutOf;
 use crate::*;
 
 /// Represent how path separator conversion should be done.
-enum Pathconversion {
+pub enum Pathconversion {
```
Comment (RalfJung) [inline src/shims/os_str.rs:17]: While you are at it, could you rename this to `PathConversion`?

**Comment by Aaron1011** [comment]
We're getting closer:

Left: `C:/Users/appveyor/AppData/Local/Temp/1/miri_test_fs_link_target.txt`
Right: `C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\/miri_test_fs_link_target.txt`

We're now returning a 'correct' path from `readlink` (a `C:` path, but with `/` separators). However, we're getting the temporary directory from the `MIRI_TEMP`, which is set by `tests/compiletest.rs` *outside* of the interpreter. I think we'll need to do path conversion when we set `MIRI_TEMP`, so that we see consistent path separators in an emulated program.

**Comment by RalfJung** [comment]
Yeah I ran into this before, where paths from the environment are the wrong way...

The thing is, the Linux target should work as well as possible on a Windows host out of the box, so we cannot necessarily expect the environment to do path conversion. But I also see no good way to do this automatically in the interpreter.

We could do something in the test (normalize separators in `MIRI_TEMP` to the current target)... if we have to adjust the test suite, I think that would be easier than adjusting `MIRI_TEMP` in `tests/compiletest.rs`.

File: tests/run-pass/fs.rs [hunk 11]
```diff
@@ -19,10 +25,23 @@ fn main() {
...
+            // program, so it may have path separators that don't
+            // correspond to our target platform. We normalize them here
+            // before constructing a `PathBuf`
+
+            #[cfg(windows)]
+            return PathBuf::from(tmp.replace("/", "\\"));
+
+            #[cfg(not(windows))]
+            return PathBuf::from(tmp.replace("\\", "/"));
+        }).unwrap_or_else(|_| std::env::temp_dir())
```
Comment (RalfJung) [inline tests/run-pass/fs.rs:44]: I start to wonder if we should put this in a lazy_static to avoid recomputing all the time.^^ But that is a question for another PR.

File: src/shims/posix/fs.rs [hunk 12]
```diff
@@ -1353,6 +1354,41 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        match result {
+            Ok(resolved) => {
+                let resolved = this.convert_path_separator(Cow::Borrowed(resolved.as_ref()), crate::shims::os_str::PathConversion::HostToTarget);
+                let mut path_bytes = crate::shims::os_str::os_str_to_bytes(resolved.as_ref())?;
+                let bufsize: usize = bufsize.try_into().unwrap();
+                if path_bytes.len() > bufsize {
+                    path_bytes = &path_bytes[..bufsize]
+                }
+                // 'readlink' truncates the resolved path if
+                // the provided buffer is not large enough
```
Comment (RalfJung) [inline src/shims/posix/fs.rs:1382]:
```suggestion
                // the provided buffer is not large enough.
```

**Comment by RalfJung** [comment]
Thanks, and nice work. :)
@bors r+
